### PR TITLE
Random sample in local mode

### PR DESCRIPTION
**Pull request by coszio**
Implements random sampling in local mode and adds the required conversions

### CODE REVIEW

File: tests/congruence_tests/test_query.py [hunk 1]
```diff
@@ -1294,3 +1308,18 @@ def test_query_group(prefer_grpc):
...
+
+    searcher = TestSimpleSearcher()
+
+    local_client = init_local()
+    init_client(local_client, fixture_points)
+
+    remote_client = init_remote(prefer_grpc=prefer_grpc)
+    init_client(remote_client, fixture_points)
+
+    compare_client_results(local_client, remote_client, searcher.random_query)
```
Comment (agourlay) [inline tests/congruence_tests/test_query.py:1325]: Why are the random implementations between local and server equivalent?
I would expect those to return different values.

File: qdrant_client/conversions/conversion.py [hunk 2]
```diff
@@ -941,6 +941,13 @@ def convert_fusion(cls, model: grpc.Fusion) -> rest.Fusion:
 
         raise ValueError(f"invalid Fusion model: {model}")  # pragma: no cover
 
+    @classmethod
+    def convert_sample(cls, model: grpc.Sample) -> rest.Sample:
+        if model == grpc.Sample.RANDOM:
```
Comment (joein) [inline qdrant_client/conversions/conversion.py:946]: same placeholder as below

File: qdrant_client/conversions/conversion.py [hunk 3]
```diff
@@ -941,6 +941,13 @@ def convert_fusion(cls, model: grpc.Fusion) -> rest.Fusion:
...
+            return rest.Sample.RANDOM
+
+        raise ValueError(f"invalid Fusion model: {model}")  # pragma: no cover
```
Comment (joein) [inline qdrant_client/conversions/conversion.py:949]:
```suggestion
        raise ValueError(f"invalid Sample model: {model}")  # pragma: no cover
```

File: tests/congruence_tests/test_query.py (hunk 1 above)
Comment (joein) [reply tests/congruence_tests/test_query.py:1325]: In this test we basically sample all the points available in the collection
In `random_query` we use `limit=100` and we generate 100 points in `fixture_points = generate_fixtures(100)`

File: qdrant_client/local/local_collection.py [hunk 4]
```diff
@@ -879,6 +878,13 @@ def _query_collection(
                 with_vectors=with_vectors,
             )
             return [record_to_scored_point(record) for record in records[offset:]]
+        elif query == models.SampleQuery(sample=models.Sample.RANDOM):
```
Comment (joein) [inline qdrant_client/local/local_collection.py:881]: just nit: seems to be a little bit odd, maybe it's better to isinstance + ==models.Sample.RANDOM?

File: qdrant_client/local/local_collection.py [hunk 5]
```diff
@@ -1767,6 +1773,49 @@ def _scroll_by_value(
...
+            payload_filter=query_filter,
+            ids_inv=self.ids_inv,
+        )
+        # in deleted: 1 - deleted, 0 - not deleted
+        # in payload_mask: 1 - accepted, 0 - rejected
+        # in mask: 1 - ok, 0 - rejected
+        mask = payload_mask & ~self.deleted
+
+        random_scores = np.random.rand(len(self.ids))
+        random_order = np.argsort(random_scores)
```
Comment (joein) [inline qdrant_client/local/local_collection.py:1794]: nit: `np.random.permutation(self.inv_ids)`

**Comment by joein** [comment]
my comments are mostly nit, other than that looks fine

**Comment by joein** [comment]
my comments are mostly nit, other than that looks good

File: tests/congruence_tests/test_query.py (hunk 1 above)
Comment (coszio) [reply tests/congruence_tests/test_query.py:1325]: We return all points postprocessed to be sorted by ID. Just to make sure we can return all of them

File: qdrant_client/proto/points.proto [hunk 6]
```diff
@@ -508,6 +508,15 @@ enum Fusion {
     DBSF = 1; // Distribution-Based Score Fusion
 }
 
+/// Sample points from the collection
+///
+/// Available sampling methods:
+///
+/// * `random` - Random sampling
+enum Sample {
+    RANDOM = 0;
```
[the remark this answers is not in the document]
Comment (coszio) [reply qdrant_client/proto/points.proto:517]: You are correct, I went along with the Fusion enum. I'll open a PR on core repo

File: qdrant_client/local/local_collection.py (hunk 5 above)
Comment (coszio) [reply qdrant_client/local/local_collection.py:1794]: but we need the internal ids to filter against the mask, not the external ones

**Review by joein — APPROVED** [review]

**Pull request merged**